Cache RMN Contract in OnRamp

asoliman92 opened the pull request:
## Motivation

Every call to arm_contract.NewARMContract deserialized ARM contract abi under the hood. Considering that OnRamp.IsSourceCursed is called very often by the background worker it creates a huge CPU overhead. Please initialize ARMContract only once and cache it instead of doing it with every method call

## Solution

* Add new cachedRmnContract field to OnRamp struct
* Cache it once per OnRamp initialization
*  fix: Use correct cache function for StaticConfig
* Add Benchmark before and after caching

Note: I did benchmark the caching function against using a map with the address as key and the function caching was a bit faster. Also using the caching is more consistent with the current implementation like `StaticConfig`.

Benchmark results:
```
BenchmarkIsSourceCursedWithCache-14        90266             12526 ns/op
BenchmarkIsSourceCursedWithCache-14        88867             12411 ns/op
BenchmarkIsSourceCursedWithCache-14        90930             12379 ns/op
BenchmarkIsSourceCursedWithCache-14        91108             12259 ns/op
BenchmarkIsSourceCursedWithCache-14        90915             12306 ns/op
BenchmarkIsSourceCursedOriginal-14          3002            397356 ns/op
BenchmarkIsSourceCursedOriginal-14          2996            402220 ns/op
BenchmarkIsSourceCursedOriginal-14          3009            399806 ns/op
BenchmarkIsSourceCursedOriginal-14          2575            397963 ns/op
BenchmarkIsSourceCursedOriginal-14          2596            396812 ns/op
```

asoliman92 committed:
Cache RMN Contract in OnRamp 1_5

* Add new cachedRmnContract field to OnRamp struct
* Cache it once per OnRamp initialization
* Add Benchmark before and after caching

asoliman92 commented on core/services/ocr2/plugins/ccip/internal/ccipdata/v1_5_0/onramp_test.go:
```diff
@@ -113,7 +113,32 @@ func Test_ProperlyRecognizesPerLaneCurses(t *testing.T) {
+// Before caching amr.NewARMContract:
+// BenchmarkIsSourceCursed-14    	    2328	    437426 ns/op
+// After caching amr.NewARMContract:
+// BenchmarkIsSourceCursed-14    	    2227	    476024 ns/op
```
This doesn't make a lot of sense!

asoliman92 commented on core/services/ocr2/plugins/ccip/internal/ccipdata/v1_5_0/onramp.go:
```diff
@@ -64,13 +64,30 @@ type OnRamp struct {
 		return nil, err
 	}
+
+	// If we are using OnceCtxFunction, and it fails, that means that the deployment won't work, What should we do in these cases?
```
This is a question I'm wondering about

0xnogo commented on core/services/ocr2/plugins/ccip/internal/ccipdata/v1_5_0/onramp.go:
```diff
@@ -64,13 +64,30 @@ type OnRamp struct {
+		return onRamp.GetStaticConfig(&bind.CallOpts{Context: ctx})
+	})
+
+	cachedRmnContract := cache.OnceCtxFunction[*arm_contract.ARMContract](func(ctx context.Context) (*arm_contract.ARMContract, error) {
```
should this be inside the `cachedStaticConfig` definition?
Might worth checking the state of `cachedStaticConfig` at this point

asoliman92 replied:
I get where you're coming from. I see most configs having simple types and I think it's better to leave it like that without complicating the config side and leave it a simple config which we can use in the app to get what we want. WDYT?

0xnogo replied:
makes sense

asoliman92 committed:
fix: Cache StaticConfig

Static config was being called with every call to `IsSourceCursed`

Needed to wrap the function call in `cache.CallOnceOnNoError` to activate the cache.

asoliman92 committed:
Add changeset

asoliman92 committed:
Cache RMN contract on onramp v1_0_0 and v1_2_0

asoliman92 committed:
Update changeset

asoliman92 committed:
Merge branch 'ccip-develop' into chore/CCIP-2391-cache-arm-contract

dimkouv commented on core/services/ocr2/plugins/ccip/internal/ccipdata/v1_5_0/onramp_test.go:
```diff
@@ -113,7 +113,32 @@ func Test_ProperlyRecognizesPerLaneCurses(t *testing.T) {
 
-func setupOnRampV1_5_0(t *testing.T, user *bind.TransactOpts, bc *client.SimulatedBackendClient) (common.Address, *mock_arm_contract.MockARMContract, common.Address) {
+// BenchmarkIsSourceCursedWithCache-14        90266             12526 ns/op
+// BenchmarkIsSourceCursedWithCache-14        88867             12411 ns/op
```
No need to keep benchmark results.

makramkd replied:
@asoliman92 you can put these in the PR description for reviewers 😄

dimkouv commented on core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/onramp.go:
```diff
@@ -86,7 +95,8 @@ func NewOnRamp(lggr logger.Logger, sourceSelector, destSelector uint64, onRampAd
 			[]common.Hash{configSetEventSig},
 			onRampAddress,
 		),
-		cachedStaticConfig: cachedStaticConfig,
```
only difference is that it's not loaded on init?

dimkouv replied:
cc @mateusz-sekara

asoliman92 replied:
> only difference is that it's not loaded on init?


Yes. When doing the benchmarks I found out that it was being called every time. Needed to wrap it within `CallOnceOnNoError` to actually cache it.

asoliman92 committed:
Merge branch 'ccip-develop' into chore/CCIP-2391-cache-arm-contract

makramkd commented on core/services/ocr2/plugins/ccip/internal/ccipdata/v1_0_0/onramp.go:
```diff
@@ -86,7 +95,8 @@ func NewOnRamp(lggr logger.Logger, sourceSelector, destSelector uint64, onRampAd
 			onRampAddress,
 		),
-		cachedStaticConfig: cachedStaticConfig,
+		cachedStaticConfig: cache.CallOnceOnNoError(cachedStaticConfig),
+		cachedRmnContract:  cache.CallOnceOnNoError(cachedRmnContract),
```
Are these called upon creating a new CCIP job? The static config one makes an rpc call, if it errors, what happens? We want to avoid a job not starting up due to intermittent rpc errors, c.f https://github.com/smartcontractkit/ccip/pull/894

asoliman92 replied:
Good question! I asked it myself when I started working on it initially as well. As you can see it's wrapped within `CallOnceOnNoError`, so in case of errors it will not actually cache. And as this is a function definition and not actually calling while starting, it is going to be called lazily when `IsSourceCursed` is called. That's why I think it shouldn't affect the job starting.

makramkd replied:
Ah I see, good note, so these aren't called here on these lines, but they will be called when IsSourceCursed is called, and only cached if there's no error. You're right, that doesn't block job startup.

asoliman92 committed:
Remove benchmark results from comments

asoliman92 committed:
Merge branch 'ccip-develop' into chore/CCIP-2391-cache-arm-contract
